mbed_targets/mbed_targets.py

```diff
@@ -58,11 +58,25 @@ def __init__(self) -> None:
             TargetDatabaseError: Failed to get the target data.
         """
         self._target_data: List[dict] = target_database.get_target_data()
+        self._target_iterator: Iterator = iter(self._target_data)
 
     def __iter__(self) -> Iterator["MbedTarget"]:
-        """Yield an MbedTarget on each iteration."""
-        for target in self._target_data:
-            yield MbedTarget(target)
+        """Return the iterator object itself.
+
+        Create a new iterator based on target data.
+        """
+        self._target_iterator = iter(self._target_data)
+        return self
+
+    def __next__(self) -> "MbedTarget":
+        """Return the next item from the iterable.
+
+        Iterate over the target data and return a new MbedTarget instance for each database entry
+
+        Returns:
+            An instance of a Mbed target database entry.
+        """
+        return MbedTarget(next(self._target_iterator))
```

Comment:
Is it desired that this re-uses `self._target_iterator` which might be reset by using `__iter__`? I rarely ever use `next` directly, so I'm not sure what this should do.

Example:
```
targets = MbedTargets()
t1 = next(targets) # has id = 1
for target in targets:
   t2 = target # has id = 1
   break
t3 = next(target) # has id = 2
```

Reply:
I think this is conceptually wrong. This `MbedTargets` class should be an iterable and not an iterator. We're actually abusing the iterator protocol here just so we can use `next` directly which is an unusual way to access elements in an iterable anyway.

Reply:
Let me provide some rationale as to why this is an abuse of the iterator protocol
* Iterators are expected to be single pass
* Iterators don't contain data, they are supposed to just implement a `__next__` method which describes how to get the next element in an iterable. Having the iterable and iterator logic in the same class violates separation of concerns and makes the code much harder to understand.
 On the other hand, iterables:
* expected to contain data and can be iterated over
* expected be multi-pass
* return a new iterator on each call to their `__iter__` method

`next` is a function designed to be used with iterators and not iterable containers

Implementing `__iter__` as  a generator (what we had before this change) fulfilled the iterable protocol; generator objects are iterators, generator functions return generator objects. This class is now trying to be both a iterable and an iterator which will actually confuse users rather than giving them what they expect, and we've now made the code a lot harder to maintain..